File: leetcode3/염혜정/3707. Equal Score Substrings.java
// 전체합을 먼저 구한 다음에 한쪽 합만 구하고 빼주면 나머지 합도 구할 수 있음
// 시간복잡도 O(n)

class Solution {
    public boolean scoreBalance(String s) {
        // a ~ b 97 ~ 122
        int total = 0;
        for (char c : s.toCharArray()) total += c - 96;
        
        int left_sum = 0;
        for (int i = 0; i<s.length()-1; i++) {
            left_sum += s.charAt(i)-96;
            int right_sum = total - left_sum;

            if (left_sum == right_sum) {
                return true;
            }
        }
        return false;
    }
}

File: leetcode3/한재원/3707. Equal Score Substrings.py
'''
input: string of alphabet letters
output: boolean - if the string can be split into substrings with equal scores

- each alphabet has the score a=1... z=26

- make dictionary key: a-z, value: 1-26
- two pointer 
        adcb
        |  |
    - while check current point of each pointer
        - if left > right
            move right pointer to the left
        - else 
            move left pointer to the right
        - return if the score is equal
    - return false

TC: O(N), SC: O(26) - O(1)

'''
class Solution:
    def scoreBalance(self, s: str) -> bool:
        score_dict = {}
        
        for i in range(26):
            score_dict[chr(ord('a') + i)] = i + 1
        
        total = 0
        
        for ch in s:
            total += score_dict[ch]
        
        left = 0
        
        for i in range(len(s) - 1):
            left += score_dict[s[i]]
            right = total - left
            
            if left == right:
                return True
        
        return False